### PR TITLE
Refactored capability parsing

### PR DESCRIPTION
**Pull request by akallabeth**
* Add new settings for previously ignored capability options
* Store raw capability data in settings for later use
* Add function to extract settings from raw capability data
* Split capability read/write from client/server logic (e.g. enforce limits, ...)

### CODE REVIEW

File: libfreerdp/common/settings.c [hunk 1]
```diff
@@ -1476,7 +1518,7 @@ void* freerdp_settings_get_pointer_array_writable(const rdpSettings* settings, s
 			max = freerdp_settings_get_uint32(settings, FreeRDP_TargetNetAddressCount);
 			if (offset >= max)
 				goto fail;
-			return &settings->TargetNetPorts[offset];
```
Comment (MartinHaimberger) [inline libfreerdp/common/settings.c:1479]: nice finding

File: libfreerdp/core/capabilities.c [hunk 2]
```diff
@@ -195,12 +180,54 @@ static BOOL rdp_read_general_capability_set(wStream* s, rdpSettings* settings)
...
+ * @return if the operation completed successfully
+ */
+
+static BOOL rdp_read_general_capability_set(wStream* s, rdpSettings* settings)
+{
+	UINT16 extraFlags;
+	BYTE refreshRectSupport;
+	BYTE suppressOutputSupport;
+
+	if (!Stream_CheckAndLogRequiredLength(TAG, s, 20))
```
Comment (MartinHaimberger) [inline libfreerdp/core/capabilities.c:206]: WINPR_ASSERT missing for settings and s

File: libfreerdp/core/capabilities.c [hunk 3]
```diff
@@ -248,15 +275,17 @@ static BOOL rdp_write_general_capability_set(wStream* s, const rdpSettings* sett
 		         settings->OsMajorType, settings->OsMinorType, UINT16_MAX);
 		return FALSE;
 	}
```
Comment (MartinHaimberger) [inline libfreerdp/core/capabilities.c:277]: WINPR_ASSERT missing on top for s and settings

File: libfreerdp/core/capabilities.c [hunk 4]
```diff
@@ -340,33 +400,15 @@ static BOOL rdp_read_bitmap_capability_set(wStream* s, rdpSettings* settings)
...
-		settings->DrawAllowSkipAlpha = (drawingFlags & DRAW_ALLOW_SKIP_ALPHA) ? TRUE : FALSE;
-
-	if (settings->DrawAllowDynamicColorFidelity)
-		settings->DrawAllowDynamicColorFidelity =
-		    (drawingFlags & DRAW_ALLOW_DYNAMIC_COLOR_FIDELITY) ? TRUE : FALSE;
-
-	if (settings->DrawAllowColorSubsampling)
-		settings->DrawAllowColorSubsampling =
-		    (drawingFlags & DRAW_ALLOW_COLOR_SUBSAMPLING) ? TRUE : FALSE;
+	freerdp_settings_set_uint32(settings, FreeRDP_ColorDepth, preferredBitsPerPixel);
```
Comment (MartinHaimberger) [inline libfreerdp/core/capabilities.c:403]: WINPR_ASSERT missing for s and settings on top

File: libfreerdp/core/capabilities.c [hunk 5]
```diff
@@ -408,12 +450,12 @@ static BOOL rdp_write_bitmap_capability_set(wStream* s, const rdpSettings* setti
 	 * appears consistent in its use.
 	 */
 
-	if ((settings->ColorDepth > UINT16_MAX) || (settings->DesktopWidth > UINT16_MAX) ||
-	    (settings->DesktopHeight > UINT16_MAX))
+	if ((freerdp_settings_get_uint32(settings, FreeRDP_ColorDepth) > UINT16_MAX) ||
```
Comment (MartinHaimberger) [inline libfreerdp/core/capabilities.c:453]: WINPR_ASSERT missing for s and settings on top

File: libfreerdp/core/capabilities.c [hunk 6]
```diff
@@ -494,56 +573,51 @@ static BOOL rdp_print_bitmap_capability_set(wStream* s)
...
-	UINT16 orderFlags;
-	BYTE orderSupport[32];
-	UINT16 orderSupportExFlags;
+	size_t i;
+	char terminalDescriptor[17] = { 0 };
+	BYTE orderSupport[32] = { 0 };
 	BOOL BitmapCacheV3Enabled = FALSE;
 	BOOL FrameMarkerCommandEnabled = FALSE;
 
+	WINPR_ASSERT(settings);
```
Comment (MartinHaimberger) [inline libfreerdp/core/capabilities.c:582]: WINPR_ASSERT missing for s

File: libfreerdp/core/capabilities.c [hunk 7]
```diff
@@ -560,20 +634,18 @@ static BOOL rdp_write_order_capability_set(wStream* s, const rdpSettings* settin
 	size_t header;
 	UINT16 orderFlags;
 	UINT16 orderSupportExFlags;
-	UINT16 textANSICodePage = 0;
+	char terminalDescriptor[16] = { 0 };
 
+	WINPR_ASSERT(settings);
```
Comment (MartinHaimberger) [inline libfreerdp/core/capabilities.c:639]: WINPR_ASSERT missing for s

File: libfreerdp/core/capabilities.c [hunk 8]
```diff
@@ -750,7 +836,7 @@ static BOOL rdp_write_bitmap_cache_capability_set(wStream* s, const rdpSettings*
 	header = rdp_capability_set_start(s);
```
Comment (MartinHaimberger) [inline libfreerdp/core/capabilities.c:836]: WINPR_ASSERT missing for s and settings

File: libfreerdp/core/capabilities.c [hunk 9]
```diff
@@ -974,7 +1090,7 @@ static BOOL rdp_read_pointer_capability_set(wStream* s, rdpSettings* settings)
 {
 	UINT16 colorPointerFlag;
 	UINT16 colorPointerCacheSize;
-	UINT16 pointerCacheSize;
+	UINT16 pointerCacheSize = 0;
 
 	if (!Stream_CheckAndLogRequiredLength(TAG, s, 4))
```
Comment (MartinHaimberger) [inline libfreerdp/core/capabilities.c:1095]: WINPR_ASSERT missing for s and settings

File: libfreerdp/core/capabilities.c [hunk 10]
```diff
@@ -1256,51 +1424,30 @@ static BOOL rdp_read_input_capability_set(wStream* s, rdpSettings* settings)
 	Stream_Read_UINT16(s, inputFlags); /* inputFlags (2 bytes) */
 	Stream_Seek_UINT16(s);             /* pad2OctetsA (2 bytes) */
```
Comment (MartinHaimberger) [inline libfreerdp/core/capabilities.c:1425]: WINPR_ASSERT missing for s and settings

File: libfreerdp/core/capabilities.c [hunk 11]
```diff
@@ -1453,7 +1617,10 @@ static BOOL rdp_print_font_capability_set(wStream* s)
 static BOOL rdp_read_brush_capability_set(wStream* s, rdpSettings* settings)
 {
 	WINPR_UNUSED(settings);
```
Comment (MartinHaimberger) [inline libfreerdp/core/capabilities.c:1619]: WINPR_ASSERT missing for s and settings

File: libfreerdp/core/capabilities.c [hunk 12]
```diff
@@ -1526,23 +1712,16 @@ static void rdp_write_cache_definition(wStream* s, GLYPH_CACHE_DEFINITION* cache
 
 static BOOL rdp_read_glyph_cache_capability_set(wStream* s, rdpSettings* settings)
 {
+	size_t x;
```
Comment (MartinHaimberger) [inline libfreerdp/core/capabilities.c:1715]: WINPR_ASSERT missing for s and settings

File: libfreerdp/core/capabilities.c [hunk 13]
```diff
@@ -1555,7 +1734,7 @@ static BOOL rdp_read_glyph_cache_capability_set(wStream* s, rdpSettings* setting
 
 static BOOL rdp_write_glyph_cache_capability_set(wStream* s, const rdpSettings* settings)
 {
-	size_t header;
+	size_t header, x;
 
 	if (!Stream_EnsureRemainingCapacity(s, 64))
```
Comment (MartinHaimberger) [inline libfreerdp/core/capabilities.c:1739]: WINPR_ASSERT missing for s and settings

File: libfreerdp/core/capabilities.c [hunk 14]
```diff
@@ -1734,10 +1925,8 @@ static BOOL rdp_read_bitmap_cache_host_support_capability_set(wStream* s, rdpSet
 	Stream_Seek_UINT8(s);               /* pad1 (1 byte) */
 	Stream_Seek_UINT16(s);              /* pad2 (2 bytes) */
 
-	if (cacheVersion & BITMAP_CACHE_V2)
-		settings->BitmapCachePersistEnabled = TRUE;
-
-	return TRUE;
+	return freerdp_settings_set_bool(settings, FreeRDP_BitmapCachePersistEnabled,
```
Comment (MartinHaimberger) [inline libfreerdp/core/capabilities.c:1928]: WINPR_ASSERT missing for s and settings

File: libfreerdp/core/capabilities.c [hunk 15]
```diff
@@ -1751,17 +1940,20 @@ static BOOL rdp_write_bitmap_cache_host_support_capability_set(wStream* s,
                                                                const rdpSettings* settings)
 {
 	size_t header;
+	UINT8 cacheVersion = 0;
+
+	if (freerdp_settings_get_bool(settings, FreeRDP_BitmapCacheEnabled))
+		cacheVersion |= BITMAP_CACHE_V2;
 
-	WINPR_UNUSED(settings);
 	if (!Stream_EnsureRemainingCapacity(s, 32))
```
Comment (MartinHaimberger) [inline libfreerdp/core/capabilities.c:1948]: WINPR_ASSERT missing for s

File: libfreerdp/core/capabilities.c [hunk 16]
```diff
@@ -1785,19 +1977,24 @@ static BOOL rdp_print_bitmap_cache_host_support_capability_set(wStream* s)
...
 	return TRUE;
 }
+#endif
 
-static void rdp_read_bitmap_cache_cell_info(wStream* s, BITMAP_CACHE_V2_CELL_INFO* cellInfo)
+static BOOL rdp_read_bitmap_cache_cell_info(wStream* s, BITMAP_CACHE_V2_CELL_INFO* cellInfo)
 {
 	UINT32 info;
+
+	if (!Stream_CheckAndLogRequiredLength(TAG, s, 4))
```
Comment (MartinHaimberger) [inline libfreerdp/core/capabilities.c:1986]: WINPR_ASSERT missing for s and cellIinfo

File: libfreerdp/core/capabilities.c [hunk 17]
```diff
@@ -1806,10 +2003,43 @@ static void rdp_write_bitmap_cache_cell_info(wStream* s, BITMAP_CACHE_V2_CELL_IN
 	 * numEntries is in the first 31 bits, while the last bit (k)
```
Comment (MartinHaimberger) [inline libfreerdp/core/capabilities.c:2003]: WINPR_ASSERT missing for s and cellIinfo

File: libfreerdp/core/capabilities.c [hunk 18]
```diff
@@ -1806,10 +2003,43 @@ static void rdp_write_bitmap_cache_cell_info(wStream* s, BITMAP_CACHE_V2_CELL_IN
...
 	 */
-	info = (cellInfo->numEntries | (((UINT32)cellInfo->persistent) << 31));
+	info = (cellInfo->numEntries | (cellInfo->persistent << 31));
 	Stream_Write_UINT32(s, info);
 }
 
+static BOOL rdp_apply_bitmap_cache_v2_capability_set(rdpSettings* settings, const rdpSettings* src)
+{
+	size_t x;
+	const size_t keys[] = { FreeRDP_BitmapCacheEnabled, FreeRDP_BitmapCachePersistEnabled };
```
Comment (MartinHaimberger) [inline libfreerdp/core/capabilities.c:2013]: WINPR_ASSERT missing for settings and src

File: libfreerdp/core/capabilities.c [hunk 19]
```diff
@@ -1820,19 +2050,31 @@ static void rdp_write_bitmap_cache_cell_info(wStream* s, BITMAP_CACHE_V2_CELL_IN
 
 static BOOL rdp_read_bitmap_cache_v2_capability_set(wStream* s, rdpSettings* settings)
 {
+	size_t x;
+	UINT16 cacheFlags = 0;
 	WINPR_UNUSED(settings);
+
```
Comment (MartinHaimberger) [inline libfreerdp/core/capabilities.c:2056]: WINPR_ASSERT missing for s

File: libfreerdp/core/capabilities.c [hunk 20]
```diff
@@ -1949,8 +2204,8 @@ static BOOL rdp_read_virtual_channel_capability_set(wStream* s, rdpSettings* set
 	else
 		VCChunkSize = 1600;
 
-	if (settings->ServerMode != TRUE)
-		settings->VirtualChannelChunkSize = VCChunkSize;
+	settings->VirtualChannelCompressionFlags = flags;
```
Comment (MartinHaimberger) [inline libfreerdp/core/capabilities.c:2207]: WINPR_ASSERT missing for settings and s on top

File: libfreerdp/core/capabilities.c [hunk 21]
```diff
@@ -1965,17 +2220,16 @@ static BOOL rdp_read_virtual_channel_capability_set(wStream* s, rdpSettings* set
 static BOOL rdp_write_virtual_channel_capability_set(wStream* s, const rdpSettings* settings)
 {
 	size_t header;
-	UINT32 flags;
 
 	if (!Stream_EnsureRemainingCapacity(s, 32))
```
Comment (MartinHaimberger) [inline libfreerdp/core/capabilities.c:2224]: WINPR_ASSERT missing for settings and s

File: libfreerdp/core/capabilities.c [hunk 22]
```diff
@@ -2022,9 +2289,8 @@ static BOOL rdp_read_draw_nine_grid_cache_capability_set(wStream* s, rdpSettings
 	Stream_Read_UINT16(s,
 	                   settings->DrawNineGridCacheEntries); /* drawNineGridCacheEntries (2 bytes) */
 
-	if ((drawNineGridSupportLevel & DRAW_NINEGRID_SUPPORTED) ||
-	    (drawNineGridSupportLevel & DRAW_NINEGRID_SUPPORTED_V2))
-		settings->DrawNineGridEnabled = TRUE;
+	settings->DrawNineGridEnabled =
```
Comment (MartinHaimberger) [inline libfreerdp/core/capabilities.c:2292]: WINPR_ASSERT missing for settings and s on top

File: libfreerdp/core/capabilities.c [hunk 23]
```diff
@@ -2125,11 +2406,8 @@ static BOOL rdp_read_draw_gdiplus_cache_capability_set(wStream* s, rdpSettings*
 	Stream_Seek(s, 8);                              /* GdipCacheChunkSize (8 bytes) */
 	Stream_Seek(s, 6);                              /* GdipImageCacheProperties (6 bytes) */
 
-	if (drawGDIPlusSupportLevel & DRAW_GDIPLUS_SUPPORTED)
-		settings->DrawGdiPlusEnabled = TRUE;
-
-	if (drawGdiplusCacheLevel & DRAW_GDIPLUS_CACHE_LEVEL_ONE)
-		settings->DrawGdiPlusCacheEnabled = TRUE;
+	settings->DrawGdiPlusEnabled = drawGDIPlusSupportLevel & DRAW_GDIPLUS_SUPPORTED;
```
Comment (MartinHaimberger) [inline libfreerdp/core/capabilities.c:2409]: WINPR_ASSERT missing for settings and s on top

File: libfreerdp/core/capabilities.c [hunk 24]
```diff
@@ -2206,23 +2505,8 @@ static BOOL rdp_read_remote_programs_capability_set(wStream* s, rdpSettings* set
...
-
-	/* 2.2.2.2.3 HandshakeEx PDU (TS_RAIL_ORDER_HANDSHAKE_EX)
-	 * the handshake ex pdu is supported when both, client and server announce
-	 * it OR if we are ready to begin enhanced remoteAPP mode. */
-	if (settings->RemoteApplicationMode)
-		railSupportLevel |= RAIL_LEVEL_HANDSHAKE_EX_SUPPORTED;
-
-	settings->RemoteApplicationSupportLevel =
-	    railSupportLevel & settings->RemoteApplicationSupportMask;
+	settings->RemoteApplicationMode = railSupportLevel & RAIL_LEVEL_SUPPORTED;
```
Comment (MartinHaimberger) [inline libfreerdp/core/capabilities.c:2508]: WINPR_ASSERT missing for settings and s on top

File: libfreerdp/core/capabilities.c [hunk 25]
```diff
@@ -2358,7 +2664,7 @@ static BOOL rdp_read_desktop_composition_capability_set(wStream* s, rdpSettings*
 	if (!Stream_CheckAndLogRequiredLength(TAG, s, 2))
```
Comment (MartinHaimberger) [inline libfreerdp/core/capabilities.c:2664]: WINPR_ASSERT missing for s

File: libfreerdp/core/capabilities.c [hunk 26]
```diff
@@ -2468,6 +2767,26 @@ static BOOL rdp_read_multifragment_update_capability_set(wStream* s, rdpSettings
...
+ * @param s stream
+ * @param settings settings
+ * @return if the operation completed successfully
+ */
+
+static BOOL rdp_read_multifragment_update_capability_set(wStream* s, rdpSettings* settings)
+{
+	UINT32 multifragMaxRequestSize;
+
+	if (!Stream_CheckAndLogRequiredLength(TAG, s, 4))
```
Comment (MartinHaimberger) [inline libfreerdp/core/capabilities.c:2785]: WINPR_ASSERT missing for settings and s

File: libfreerdp/core/capabilities.c [hunk 27]
```diff
@@ -2778,7 +3141,7 @@ static BOOL rdp_read_bitmap_codecs_capability_set(wStream* s, rdpSettings* setti
 		if (!Stream_SafeSeek(s, codecPropertiesLength))
 			return FALSE;
 
-		if (settings->ServerMode)
+		if (isServer)
```
Comment (MartinHaimberger) [inline libfreerdp/core/capabilities.c:3144]: WINPR_ASSERT missing for settings and s on top

File: libfreerdp/core/capabilities.c [hunk 28]
```diff
@@ -3259,14 +3630,7 @@ static BOOL rdp_read_frame_acknowledge_capability_set(wStream* s, rdpSettings* s
...
 
-	if (settings->ServerMode)
-	{
-		Stream_Read_UINT32(s, settings->FrameAcknowledge); /* (4 bytes) */
-	}
-	else
-	{
-		Stream_Seek_UINT32(s); /* (4 bytes) */
-	}
+	Stream_Read_UINT32(s, settings->FrameAcknowledge); /* (4 bytes) */
```
Comment (MartinHaimberger) [inline libfreerdp/core/capabilities.c:3633]: WINPR_ASSERT missing for settings and s

File: libfreerdp/core/capabilities.c [hunk 29]
```diff
@@ -3316,6 +3690,7 @@ static BOOL rdp_read_bitmap_cache_v3_codec_id_capability_set(wStream* s, rdpSett
 		return FALSE;
```
Comment (MartinHaimberger) [inline libfreerdp/core/capabilities.c:3690]: WINPR_ASSERT missing for s

File: libfreerdp/core/capabilities.c [hunk 30]
```diff
@@ -3589,265 +3964,366 @@ BOOL rdp_print_capability_sets(wStream* s, size_t start, BOOL receiving)
...
-	UINT16 numberCapabilities;
-	UINT16 count;
+	size_t rest;
 
-#ifdef WITH_DEBUG_CAPABILITIES
-	const size_t capstart = Stream_GetPosition(s);
-#endif
-	if (!Stream_CheckAndLogRequiredLength(TAG, s, 4))
-		return FALSE;
+	if (type <= CAPSET_TYPE_FRAME_ACKNOWLEDGE)
```
Comment (MartinHaimberger) [inline libfreerdp/core/capabilities.c:4039]: WINPR_ASSERT missing for settings and sub

File: libfreerdp/core/capabilities.c [hunk 31]
```diff
@@ -4030,7 +4511,7 @@ static BOOL rdp_write_demand_active(wStream* s, rdpSettings* settings)
 		return FALSE;
 	}
 
-	if (settings->BitmapCachePersistEnabled)
+	if (freerdp_settings_get_bool(settings, FreeRDP_BitmapCachePersistEnabled))
```
Comment (MartinHaimberger) [inline libfreerdp/core/capabilities.c:4514]: WINPR_ASSERT missing for settings and s on top

File: libfreerdp/core/capabilities.c [hunk 32]
```diff
@@ -113,7 +113,6 @@ static BOOL rdp_read_capability_set_header(wStream* s, UINT16* length, UINT16* t
```
Comment (MartinHaimberger) [inline libfreerdp/core/capabilities.c:113]: Not all functions where touched, but ensure that all get and set methods are using WINPR_ASSERT before accessing a pointer, check the whole file!

**Review by MartinHaimberger — CHANGES_REQUESTED** [review]
Check each function in capabilities for correct ASSERT

File: libfreerdp/core/capabilities.c (hunk 2 above)
Comment (akallabeth) [reply libfreerdp/core/capabilities.c:206]: `s` is asserted by `Stream_CheckAndLogRequiredLength`

**Comment by akallabeth** [comment]
@MartinHaimberger added `WINPR_ASSERT` for `settings`, `wStream*` is already asserted by all stream functions, so not required.

**Review by MartinHaimberger — APPROVED** [review]
LGTM

**Pull request closed without merging**